### PR TITLE
v2.0 rollout plan draft

### PR DESCRIPTION
**Pull request by neilmb**
This is a draft document describing the technical plan for the CEJST v2.0 release to the public.

@ericiwamoto Can you review this and either leave comments or send an email with your thoughts? @vim-usds you know as much as anyone about the architecture here if you also want to take a look.

### CODE REVIEW

File: docs/operations/rollout-plan.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+The frontend application uses that URL and some path configuration information
+to provide the correct data.
+
+The frontend application is a static site built by Gatsby in a Github Action
+then deployed to and served from an S3 bucket behind a Cloudfront distribution at
+screeningtool.geoplatform.gov.
+
+## Backend Data Rollout
+
+The CEJST data is stored in an S3 bucket under "versioned" paths (we think)
```
Comment (vim-usds) [inline docs/operations/rollout-plan.md:22]: we "should" 🤞🏾  be able to simply update this [env](https://github.com/DOI-DO/ceq-j40-cejst-2/blob/8712ff53ec07107ba60412f1e10d7957a7f865b4/client/.env.development#L13) var to point to the new 2.0 location for the map tiles (I think).

File: docs/operations/rollout-plan.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+Rather than trying to incrementally change the frontend app that is deployed
+at screeningtool.geoplatform.gov, we will deploy the new 2.0 version of the
+frontend website in a new S3 bucket with a new Cloudfront distribution, and
+configured to use the new backend data under the `2.0/` paths. This
+will allow us to test the deploy process, website improvements, and the
+display of the updated data using a staging URL like
+<random-string-of-letters>.cloudfront.net.
+
+TODO: Get new AWS resources
+  - A new website S3 bucket
```
Comment (vim-usds) [inline docs/operations/rollout-plan.md:46]: Totally agreed on a new 2.0 URL

**Review by vim-usds — APPROVED** [review]
Methodical approach the limits risks, enables experimentation while working out kinks of 2.0 all while not having any chance of disrupting the current 1.0 production site. Really well thought out. TY ❤️

**Comment by vim-usds** [comment]
@neilmb and @KameronKerger :

This PR outlines the fastest way to move forward. Period. 

While I believe it may be at a cost (potentially) to long term stability. I'm not sure if the following can or should be addressed now. 

### Frontend
1. Moving away from Gatsby. Gatsby was the hot SSG in 2021, but last I checked it is no longer getting updates and is not actively being developed. This is a risk in IMHO.
2. Moving to Vite / Astro / Next.js as a more modern solution

### Backend
The way CEJST 1.0 map tile creation / updating was originally designed ends up taking a long time to test changes to the pipeline on the map. There are probably smarter ways to 1. create the map tiles, 2. update the map tiles and 3 serve the map tiles. Given that iterating on them takes so long, it makes development and experimentation painful. If I had unlimited resources and time, I would take a step back and revisit this architecture. If a full redo on it wasn't possible, are there low hanging fruits we can do to make this experience better? Our old DS Emma Nechamkin had ideas on ways this could be made better by using a geo DB (using Postgres). This implies adding a backend layer and possibly EC2 or Lambdas. An ADR would be needed to evaluate the tradeoffs. I think this could be worthwhile.

The above to suggestions are hinting at a larger thought: It feels like the CEQ products tend to be in "fire fighting mode". What I mean by this is - we have a small amount of time to get something out, people work very fast, take shortcuts to get things out. For the short term this is OK, but also could have a tendency to lead (potentially) burnout, loosing good people (that care) and maybe isn't sustainable. Can we afford to invest into a sustainable PDLC that doesn't require that? 🤷🏾‍♂️ 

Just my .02 cents. Thanks for entertaining.

**Comment by KameronKerger** [comment]
Thanks for this. We might have runway in Jan to look at making these fixes!
Sent from my iPhone

On Dec 4, 2024, at 11:42 AM, Vim ***@***.***> wrote:

﻿

@neilmb<https://github.com/neilmb> and @KameronKerger<https://github.com/KameronKerger> :

This PR outlines the fastest way to move forward. Period.

While I believe it may be at a cost (potentially) to long term stability. I'm not sure if the following can or should be addressed now.

Frontend

  1.  Moving away from Gatsby. Gatsby was the hot SSG in 2021, but last I checked it is no longer getting updates and is not actively being developed. This is a risk in IMHO.
  2.  Moving to Vite / Astro / Next.js as a more modern solution

Backend

The way CEJST 1.0 map tile creation / updating was originally designed ends up taking a long time to test changes to the pipeline on the map. There are probably smarter ways to 1. create the map tiles, 2. update the map tiles and 3 serve the map tiles. Given that iterating on them takes so long, it makes development and experimentation painful. If I had unlimited resources and time, I would take a step back and revisit this architecture. If a full redo on it wasn't possible, are there low hanging fruits we can do to make this experience better? Our old DS Emma Nechamkin had ideas on ways this could be made better by using a geo DB (using Postgres). This implies adding a backend layer and possibly EC2 or Lambdas. An ADR would be needed to evaluate the tradeoffs. I think this could be worthwhile.

The above to suggestions are hinting at a larger thought: It feels like the CEQ products tend to be in "fire fighting mode". What I mean by this is - we have a small amount of time to get something out, people work very fast, take shortcuts to get things out. For the short term this is OK, but also could have a tendency to lead (potentially) burnout, loosing good people (that care) and maybe isn't sustainable. Can we afford to invest into a sustainable PDLC that doesn't require that? 🤷🏾‍♂️

Just my .02 cents. Thanks for entertaining.

File: docs/operations/rollout-plan.md (hunk 1 above)
Comment (ericiwamoto) [reply docs/operations/rollout-plan.md:22]: This is correct. We built a test environment in Agile6's AWS account that should be a close replica of the existing production environment, and we updated that exact env var to point to the v2.0 map tiles we deployed.

**Comment by ericiwamoto** [comment]
@neilmb The proposal looks good to me, as it covers everything we discussed and should allow us to get 2.0 deployed quickly while minimizing risk to the existing site.

**Pull request merged**